### PR TITLE
[Enhancement] percentile_cont performance enhance

### PR DESCRIPTION
**Pull request by XiaoLawn**
Why I'm doing:
percentile_cont performance enhance

What I'm doing:
Detailed information about this enhancement has been put in issue below:
https://github.com/StarRocks/starrocks/issues/36160

Test on the same cluster and the same data set, and record the performance before and after optimization.

8506500 items —— 0.54 sec --> 0.20 sec

54387164 items —— 5.15 sec --> 0.80 sec

109610515 items —— 8.65 sec --> 1.50 sec

251414166 items —— 20.10 sec --> 3.42 sec

742991422 items —— 1 min 14.80 sec --> 10.88 sec

1106646823 items —— 2 min 55.81 sec --> 16.16 sec

The test shows that the processing performance has been significantly improved.
Especially the processing performance under large data volume is close to 10 times that before optimization.

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/test/exprs/agg/aggregate_test.cpp [hunk 1]
```diff
@@ -1394,6 +1394,63 @@ TEST_F(AggregateTest, test_percentile_cont) {
     ASSERT_EQ(3, result_column->get_data()[0]);
 }
 
+TEST_F(AggregateTest, test_percentile_cont_non_empty_grid_serialize) {
```
Comment (imay) [inline be/test/exprs/agg/aggregate_test.cpp:1397]: why not use sql-test to add more tests? That way is much easier.

**Review by imay — COMMENTED** [review]
Could you paste the performance gain from this change?

**Comment by rubiesvelt** [comment]
> @rubiesvelt it would be appreciated if you can open an issue, and move the comprehensive context and background information to the issue, just keep a few brief summary here and give a reference to the issue.

Sure, detailed information about this enhancement has been put in the issue https://github.com/StarRocks/starrocks/issues/36160

**Comment by rubiesvelt** [comment]
> Could you paste the performance gain from this change?

No problem, performance gain has been put in the PR and detailed information about this enhancement had been put in this issue https://github.com/StarRocks/starrocks/issues/36160

File: be/test/exprs/agg/aggregate_test.cpp (hunk 1 above)
Comment (rubiesvelt) [reply be/test/exprs/agg/aggregate_test.cpp]: > why not use sql-test to add more tests? That way is much easier.

Sorry, I just failed to find the usage of sql-test, Could you please give me an example?

Comment (rubiesvelt) [reply be/test/exprs/agg/aggregate_test.cpp]: > > why not use sql-test to add more tests? That way is much easier.
> 
> Sorry, I just failed to find the usage of sql-test, Could you please give me an example?

Already founded, never mind

**Review by satanson — DISMISSED** [review]

File: be/src/exprs/agg/percentile_cont.h [hunk 2]
```diff
@@ -92,18 +208,29 @@ class PercentileContDiscAggregateFunction
...
-        pdqsort(reinterpret_cast<InputCppType*>(bytes.data() + old_size + sizeof(double) + sizeof(size_t)),
-                reinterpret_cast<InputCppType*>(bytes.data() + old_size + sizeof(double) + sizeof(size_t) +
-                                                items_size * sizeof(InputCppType)));
+        size_t current_pos = old_size + sizeof(double) + sizeof(size_t) + items_size * sizeof(InputCppType);
+        for (const auto& vec : this->data(state).grid) {
+            memcpy(bytes.data() + current_pos, vec.data() + 1, (vec.size() - 2) * sizeof(InputCppType));
+            current_pos += (vec.size() - 2) * sizeof(InputCppType);
+        }
+
+        std::sort(reinterpret_cast<InputCppType*>(bytes.data() + old_size + sizeof(double) + sizeof(size_t)),
```
Comment (mofeiatwork) [inline be/src/exprs/agg/percentile_cont.h:231]: why not use `pdqsort` ? it's faster than `std::sort`

File: be/src/exprs/agg/percentile_cont.h [hunk 3]
```diff
@@ -40,11 +40,120 @@ template <LogicalType LT, typename = guard::Guard>
...
+                std::swap(q, ls[t]);
+            }
+            t = t / 2;
+        }
+        ls[0] = q;
+    }
+}
+
+template <LogicalType LT, typename CppType, typename ResultType>
+ResultType calculateResult(CppType juniorElm, CppType seniorElm, double u, size_t index) {
```
Comment (mofeiatwork) [inline be/src/exprs/agg/percentile_cont.h:141]: naming

File: be/src/exprs/agg/percentile_cont.h [hunk 4]
```diff
@@ -40,11 +40,120 @@ template <LogicalType LT, typename = guard::Guard>
...
+    if constexpr (lt_is_datetime<LT>) {
+        result.from_unix_second(juniorElm.to_unix_second() +
+                                (u - (float)index) * (seniorElm.to_unix_second() - juniorElm.to_unix_second()));
+    } else if constexpr (lt_is_date<LT>) {
+        result._julian = juniorElm._julian + (u - (double)index) * (seniorElm._julian - juniorElm._julian);
+    } else if constexpr (lt_is_arithmetic<LT>) {
+        result = juniorElm + (u - (double)index) * (seniorElm - juniorElm);
+    } else {
+        // won't go there if percentile_cont is registered correctly
+        throw std::runtime_error("Invalid LogicalTypes for percentile_cont function");
```
Comment (mofeiatwork) [inline be/src/exprs/agg/percentile_cont.h:152]: Please report the specific exceptional type name

**Review by murphyatwork — COMMENTED** [review]
did you compare the performance with `starrocks::merge_sorted_chunks` ?

File: be/src/exprs/agg/percentile_cont.h (hunk 4 above)
Comment (rubiesvelt) [reply be/src/exprs/agg/percentile_cont.h]: Sure, already added

File: be/src/exprs/agg/percentile_cont.h (hunk 3 above)
Comment (rubiesvelt) [reply be/src/exprs/agg/percentile_cont.h]: Naming fixed, thanks

File: be/src/exprs/agg/percentile_cont.h (hunk 2 above)
Comment (rubiesvelt) [reply be/src/exprs/agg/percentile_cont.h]: Already modified, thanks a lot

**Comment by rubiesvelt** [comment]
> did you compare the performance with `starrocks::merge_sorted_chunks` ?

No we didn't do that comparison yet, we tried to optimize the performance to the best and did comparison with the previous approach.

**Review by murphyatwork — APPROVED** [review]

**Review by satanson — APPROVED** [review]

**Pull request merged**